### ISSUE TITLE
[sdb] Impossible to step through multithreaded code

### ISSUE THREAD

**Issue by jbevain** [mono/mono#14456]
Let's have a look at the following code:

```csharp
void Start()
{
   for (int i = 0; i < 5; i++)
       Thread.Start(Foo);
}

void Foo()
{
  int a = 12;
  int b = 13;
  int c = 13;
}
```

Using `Mono.Debugger.Soft`, you put a breakpoint at the line that reads: `int a = 12;`

Upon executing the code, you'll hit the breakpoint on the new thread `0`.
You create a step request for this thread.
You resume.

You'll hit the breakpoint on the new thread `1`.
You create a step request for this thread.

You get a ERROR_NOT_IMPL from the server:

https://github.com/mono/mono/blob/b0ae9dbdaac492e1a047595c7774ef499658d40a/mono/mini/debugger-engine.c#L1488

Basically, this forces a design where there's a single stepping request instance globally, making it impossible to step through multiple threads.

**Comment by jbevain** [comment]
@marek-safar @thaystg @joncham that would be fantastic to have before Unity takes a newer Mono.

**Comment by thaystg** [comment]
Okay, I will try to come back to this implementation on next sprint :)
I'll let you know.

**Comment by vargaz** [comment]
Is this behavior different than how vs/gdb etc. work ?

**Comment by thaystg** [comment]
I don't know if I understood the question, but if you insert a breakpoint at this line: `int a = 12;` when running on VSWin for example, you can stop on each thread in this line, and to step over or step out or continue for each thread. Using mono with sdb or vs, we just stop on all breakpoints, but only the single step of last thread that you SS works.

**Comment by vargaz** [comment]
When the user presses the single step button, the ide should cancel the previous single step req if any, then issue a single step for that thread. Wouldn't that work ?

**Comment by thaystg** [comment]
This works but this is exactly what we don't want, to want mono to handle with single step for 2 or more threads at same time.
